x-pack/legacy/plugins/spaces/server/lib/copy_to_spaces/copy_to_spaces.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract, SavedObjectsService, SavedObject } from 'src/core/server';
+import { Readable } from 'stream';
+import { SavedObjectsClientProviderOptions } from 'src/core/server';
+import { spaceIdToNamespace } from '../utils/namespace';
+import { CopyOptions, CopyResponse } from './types';
+import { getEligibleTypes } from './lib/get_eligible_types';
+import { createEmptyFailureResponse } from './lib/create_empty_failure_response';
+import { readStreamToCompletion } from './lib/read_stream_to_completion';
+
+export const COPY_TO_SPACES_SAVED_OBJECTS_CLIENT_OPTS: SavedObjectsClientProviderOptions = {
+  excludedWrappers: ['spaces'],
+};
+
+export function copySavedObjectsToSpacesFactory(
+  savedObjectsClient: SavedObjectsClientContract,
+  savedObjectsService: SavedObjectsService
+) {
+  const { importExport, types, schema } = savedObjectsService;
+  const eligibleTypes = getEligibleTypes({ types, schema });
+
+  const exportRequestedObjects = async (
+    sourceSpaceId: string,
+    options: Pick<CopyOptions, 'includeReferences' | 'objects'>
+  ) => {
+    const objectStream = await importExport.getSortedObjectsForExport({
+      namespace: spaceIdToNamespace(sourceSpaceId),
+      includeReferencesDeep: options.includeReferences,
+      objects: options.objects,
+      savedObjectsClient,
+      types: eligibleTypes,
+      exportSizeLimit: importExport.objectLimit,
+    });
+
+    return readStreamToCompletion<SavedObject>(objectStream);
+  };
+
+  const importObjectsToSpace = async (
+    spaceId: string,
+    objectsStream: Readable,
+    options: CopyOptions
+  ) => {
+    try {
+      const importResponse = await importExport.importSavedObjects({
+        namespace: spaceIdToNamespace(spaceId),
+        objectLimit: importExport.objectLimit,
+        overwrite: options.overwrite,
+        savedObjectsClient,
+        supportedTypes: eligibleTypes,
+        readStream: objectsStream,
+      });
+
+      return {
+        success: importResponse.success,
+        successCount: importResponse.successCount,
+        errors: importResponse.errors,
+      };
+    } catch (error) {
+      return createEmptyFailureResponse([error]);
+    }
+  };
+
+  const copySavedObjectsToSpaces = async (
+    sourceSpaceId: string,
+    destinationSpaceIds: string[],
+    options: CopyOptions
+  ): Promise<CopyResponse> => {
+    const response: CopyResponse = {};
+
+    const exportedSavedObjects = await exportRequestedObjects(sourceSpaceId, options);
+
+    for (const spaceId of destinationSpaceIds) {
+      response[spaceId] = await importObjectsToSpace(
+        spaceId,
+        new Readable({
```

Comment:
It seems like we're doing this a few places, should we use `createListStream` from `../../../../../../../src/legacy/utils/streams` here? `createListStream` doesn't implement the "back-pressure" properly, but we should likely just fix this in general.

Reply:
I considered using the legacy stream utils here, but I didn't because it looks like there are plans to [deprecate/replace these](https://github.com/elastic/kibana/pull/39674#discussion_r298977279) with other stable packages as part of the NP migration.

What if I pull this out into a utility within the [`copy_to_spaces/lib` directory](https://github.com/legrego/kibana/tree/spaces/copy-to-space-api/x-pack/legacy/plugins/spaces/server/lib/copy_to_spaces/lib) with the other helpers instead?

I don't have a strong opinion either way here, so I'll happily do whatever.

Reply:
Ah, gotcha. Adding a separate utility seems fine then, especially since once we upgrade NodeJS we can just use `Readable.from`